micronota/bfillings/prodigal.py

```diff
@@ -137,6 +144,144 @@ def predict_genes(in_fp, out_dir, prefix, params=None):
         params[i] = out_fp
 
     params['-i'] = in_fp
-
+    if '-p' not in params:
+        params['-p'] = 'meta'
     app = Prodigal(params=params)
     return app()
+
+
+def predict_cds(seq, out_dir, prefix=None, params=None):
+    '''Predict genes for the input sequence with ``Prodigal``.
+
+    Parameters
+    ----------
+    seq : str or ``skbio.Sequence`` or its child classes.
+        input fasta
+    out_dir : str
+        output directory
+
+    Returns
+    -------
+    ``skbio.Sequence`` or its child classes.
+    '''
+    if isinstance(seq, str):
+        seq = Sequence(seq, metadata={'id': id(seq)})
+    if not isinstance(seq, Sequence):
+        raise
+    with NamedTemporaryFile(mode='w+') as f:
+        seq.write(f)
+        interval_metadata = {}
+        if prefix is None:
+            prefix = seq.metadata['id']
+        res = identify_features(f.name, out_dir, prefix, params)
+        if res['Exitstatus'] != 0:
+            raise RuntimeError(
+                'The Prodigal prediction is finished with an error:\n%s' %
+                res['StdErr'])
+        interval_metadata = next(parse_output(res))
+        seq.interval_metadata = interval_metadata
+        return seq
+
+
+def parse_output(res):
+    '''Parse gene prediction result from ``Prodigal``.
+
+    It is parsed into a dict consumable by``skbio.meta.IntervalMetadata``.
+
+    Parameters
+    ----------
+    fh : file handler
+
+    Returns
+    -------
+    ``IntervalMetadata``
+    '''
+    # make sure to move to the beginning of the file.
+    # return _parse_records(res['-o'], _parse_single_record)
+    return _parse_faa(res['-a'])
+
+
+def _parse_single_record(chunks):
+    '''Parse single record of Prodigal GenBank output.
+
+    Parameters
+    ----------
+    chunks : list of str
+        a list of lines of the record to parse.
+
+    Returns
+    -------
+    ``skbio.metadata.IntervalMetadata``
+    '''
+    # get the head line
+    head = chunks[0]
+    _, description = head.split(None, 1)
+    pattern = re.compile(r'''((?:[^;"']|"[^"]*"|'[^']*')+)''')
+    desc = {}
+    for i in pattern.findall(description):
+        k, v = i.split('=', 1)
+        desc[k] = v
+    return _parse_features(chunks[1:], int(desc['seqlen']))
+
+
+def _parse_faa(faa):
+    '''Parse the faa output of Prodigal.
+
+    Notice that Prodigal do not predict genes with introns.
+
+    Returns
+    -------
+    The dict passable to ``skbio.metadata.IntervalMetadata``.
+    '''
+    im = dict()
+    i = 1
+    for seq in read(faa, format='fasta'):
+        desc = seq.metadata['description']
+        pattern = (r'# +([0-9]+)'    # start
+                   ' +# +([0-9]+)'   # end
+                   ' +# +(-?1)'      # strand
+                   ' +# +ID=([0-9]+_[0-9]+);'
+                   'partial=([01]{2});'
+                   '(.*)')
```

Comment:
Can we move this pattern outside of the for loop? No point in repeatedly initializing it.